### PR TITLE
Feature/common random manyrun

### PR DESCRIPTION
**Pull request by adolgert**
The CommonRandomNumber struct wasn't conceptually sound. This time, it's set up so that you:

 1. Run the simulation many times to record random number generator states.
 2. Then use those stored values to run subsequent measurements.

Before, it assumed, quite incorrectly, that you would run the simulation once, record values, and move on. The change in the code isn't big. I added a reset! method to tell the CRN that we are starting a fresh run of the simulation. That's it. But it becomes a reset! method on every sampler, so I added it to the sampler interface. And added more testing, especially for the keyed prefix sum.

### CODE REVIEW

File: docs/src/commonrandom.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  4. For every parameter set to try, run it the same way, using `reset!` after each run.
+  5. Compare outcomes.
+
+Because the `CommonRandomRecorder` stores the state of the random number generator at each step, it works best with random number generators that have small state, such as Xoshiro on a linear congruential generator (LCG).
+
+```julia
+using Random: Xoshiro
+using Fleck
+example_clock = (3, 7)  # We will use clock IDs that are a tuple of 2 integers.
+sampler = FirstToFire{typeof(example_clock)}()
```
Comment (slwu89) [inline docs/src/commonrandom.md:25]: Needs the 2nd type parameter, for the time type. Maybe a good reason to produce these from .jl files?

File: docs/src/commonrandom.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+What we want is a variance reduction technique. Common random numbers (CRN) are a variance reduction technique that enables you to use fewer simulation runs to compare the effect of different simulation parameters on the outcome. There are several other variance reduction techniques, such as antithetic variates and importance sampling, but let's look at common random numbers in Fleck.
+
+## Using Common Random Numbers in Fleck
+
+Fleck implements common random numbers by recording the state of the random number generator every time a clock is enabled. There are other ways to do this, but this one works with the [CombinedNextReaction](@ref) and [FirstToFire](@ref) samplers. The workflow you would use looks notionally like:
+
+  1. Create a sampler.
+  2. Wrap it in a [CommonRandomRecorder](@ref).
+  3. Run a lot of simulations in order to explore and record all possible clock states. Run `reset!(recorder)` after each simulation.
```
Comment (slwu89) [inline docs/src/commonrandom.md:15]: Can we say more about why this is necessary? A reader might reasonably wonder if this extra simulation time for this step would outweigh the saved simulation time from doing fewer runs during estimation.

**Review by slwu89 — COMMENTED** [review]
The text in commonrandom.md is clear. I wonder if we could also include a dead simple example of using this? Maybe evaluating performance of a M/M/1 queue and showing reduction in variance using CRN vs crude monte carlo?

File: docs/src/commonrandom.md (hunk 1 above)
Comment (adolgert) [reply docs/src/commonrandom.md:25]: Ah, you're right. As long as we are working with Literate.jl, I guess we might as well continue.

**Pull request merged**